pkg/app/ops/insightcollector/collector.go

```diff
@@ -13,3 +13,361 @@
 // limitations under the License.
 
 package insightcollector
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/pipe-cd/pipe/pkg/filestore"
+
+	"github.com/pipe-cd/pipe/pkg/insightstore"
+
+	"go.uber.org/zap"
+
+	"github.com/pipe-cd/pipe/pkg/datastore"
+	"github.com/pipe-cd/pipe/pkg/model"
+)
+
+var aggregateKinds = []model.InsightMetricsKind{
+	model.InsightMetricsKind_DEPLOYMENT_FREQUENCY,
+	model.InsightMetricsKind_CHANGE_FAILURE_RATE,
+}
+
+// InsightCollector implements the behaviors for the gRPC definitions of InsightCollector.
+type InsightCollector struct {
+	applicationStore datastore.ApplicationStore
+	deploymentStore  datastore.DeploymentStore
+	insightstore     insightstore.Store
+	logger           *zap.Logger
+}
+
+// NewInsightCollector creates a new InsightCollector instance.
+func NewInsightCollector(
+	ds datastore.DataStore,
+	fs filestore.Store,
+	logger *zap.Logger) *InsightCollector {
+	a := &InsightCollector{
+		applicationStore: datastore.NewApplicationStore(ds),
+		deploymentStore:  datastore.NewDeploymentStore(ds),
+		insightstore:     insightstore.NewStore(fs),
+		logger:           logger.Named("insight-collector"),
+	}
+	return a
+}
+
+func (i *InsightCollector) Run(ctx context.Context) error {
+	now := time.Now().UTC()
+	maxUpdateAt := now.Unix()
+
+	for {
```

Comment:
I don't get the implementation of `robfig/cron` in detail, so please let me know why we should perform an infinite loop 🙏

Reply:
This loop is not related to robfig/cron.

The filter is set to fetch from the most recently updated applications, and the loop is [breaked](https://github.com/pipe-cd/pipe/pull/1287/files#diff-5ae729fb58b60f7289123947f3197ac5707ee720e5420d36ae0d045f3b1febd1R87-R90) when all applications have been fetched.

Reply:
Got it. You plan to increment the page after resolving https://github.com/pipe-cd/pipe/issues/1195, right?

Reply:
Yes, right

Reply:
I think the solution used in this PR is not depending on #1195. 
This is using the `createdAt` field to sort and list the applications.
And we don't need the added TODO too.

Reply:
Okay, now I found I was completely wrong.